src/main/java/org/broadinstitute/consent/http/service/dao/DatasetServiceDAO.java

```diff
@@ -0,0 +1,171 @@
+                    // Turn that off for this connection. This will not affect existing or new connections and
+                    // only applies to the current one in this handle.
+                    handle.getConnection().setAutoCommit(false);
+                    // 1. Generate inserts for missing dictionary terms
+                    // 2. Generate inserts for new dataset properties
+                    // 3. Generate updates for existing dataset properties
+                    // 4. Generate deletes for outdated dataset properties
+                    List<Update> updates = new ArrayList<>(generateDictionaryInserts(handle, properties));
+                    // We need to know existing properties for all property operations
+                    Set<DatasetProperty> existingProps = datasetDAO.findDatasetPropertiesByDatasetId(datasetId);
+                    updates.addAll(generatePropertyInserts(handle, datasetId, properties, existingProps));
+                    updates.addAll(generatePropertyUpdates(handle, datasetId, properties, existingProps));
+                    updates.addAll(generatePropertyDeletes(handle, properties, existingProps));
+                    updates.forEach(Update::execute);
+                    handle.commit();
+                }
+        );
+        return datasetDAO.findDatasetPropertiesByDatasetId(datasetId).stream().toList();
+    }
+
+    // Helper methods to generate Dictionary inserts
+
+    private List<Update> generateDictionaryInserts(Handle handle, List<DatasetProperty> properties) {
+        List<Dictionary> dictionaryTerms = datasetDAO.getDictionaryTerms();
+        List<String> keyValues = dictionaryTerms.stream().map(Dictionary::getKey).toList();
+        List<Update> updates = new ArrayList<>();
+        properties.forEach(prop -> {
+            if (!keyValues.contains(prop.getPropertyName())) {
+                updates.add(createDictionaryInsert(handle, prop.getPropertyName()));
+            }
+        });
+        return updates;
+    }
+
+    private Update createDictionaryInsert(Handle handle, String key) {
+        final String sql = """
+            INSERT INTO dictionary (key, required)
+            VALUES (:key, FALSE)
+            ON CONFLICT DO NOTHING
+        """;
+        Update insert = handle.createUpdate(sql);
+        insert.bind("key", key);
+        return insert;
+    }
+
+    // Helper methods to generate DatasetProperty inserts
+
+    private List<Update> generatePropertyInserts(Handle handle, Integer datasetId, List<DatasetProperty> properties, Set<DatasetProperty> existingProps) {
+        Timestamp now = new Timestamp(new Date().getTime());
+        List<Update> updates = new ArrayList<>();
+        List<String> existingPropNames = existingProps.stream().map(DatasetProperty::getPropertyName).toList();
+        // Generate new inserts for props we don't know about yet
+        properties.forEach(prop -> {
+            if (!existingPropNames.contains(prop.getPropertyName())) {
```

Comment:
If you're checking to see if a name is included in a list, it'll be better to use a `HashSet` (which has an `O(1)` lookup) instead of a `ArrayList` (which has an `O(length of array)` lookup).

Reply:
Yes, I was reading up on this for one of Connor's PRs ... will do.